tests/Unit/PrestaShopBundle/Translation/TranslationCatalogueBuilderTest.php

```diff
@@ -0,0 +1,482 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace Tests\Unit\PrestaShopBundle\Translation;
+
+use PHPUnit\Framework\TestCase;
+use PrestaShopBundle\Translation\Exception\UnexpectedTranslationTypeException;
+use PrestaShopBundle\Translation\Provider\CatalogueProviderFactory;
+use PrestaShopBundle\Translation\Provider\CatalogueProviderInterface;
+use PrestaShopBundle\Translation\TranslationCatalogueBuilder;
+use PrestaShopBundle\Translation\View\TranslationsTreeBuilder;
+use Symfony\Bundle\FrameworkBundle\Routing\Router;
+use Symfony\Component\Translation\MessageCatalogue;
+
+class TranslationCatalogueBuilderTest extends TestCase
+{
+    private const LOCALE = 'en-US';
+
+    private static $defaultTranslations = [
+        'AdminFirstDomain' => [
+            'First Domain First Wording' => 'First Domain First Wording Default Translation',
+            'First Domain Second Wording' => 'First Domain Second Wording Default Translation',
+        ],
+        'AdminSecondDomain' => [
+            'Second Domain First Wording' => 'Second Domain First Wording Default Translation',
+            'Second Domain Second Wording' => 'Second Domain Second Wording Default Translation',
+        ],
+    ];
+    private static $fileTranslatedTranslations = [
```

Comment:
Maybe we could have multiple sets of data for testing.
One simple translation catalogue (this one) with two or three domains.
-> this one is good to validate simple and logic operations

But we would also need more complex usecases. Maybe we could actually use a real Translations catalog from PrestaShop? And assert what should be built from it?

However as PHPUnit Assertions are verbose, the test might become unreadable...
I think we need to build a few helper functions that allow us to easily write tests like "I give you the input catalog, I give you the expected output, please validate it". Something like a Tree Crawler or an Iterator?

ping @jolelievre I'm sure you'll have insights